### ISSUE TITLE
Make kube-proxy ipset parameters 'hashSize' and 'maxElem' customizable

### ISSUE THREAD

**Issue by weak-fox** [kubernetes/kubernetes#129389]
### What would you like to be added?

Allow users to configure ipset parameters 'hashSize' and 'maxElem' in kube-proxy configuration. These parameters should be customizable via the kube-proxy config file or command line flags.

### Why is this needed?

In our use case, we have multiple LAN devices that need to interact with services in the cluster via RTP media streams. Each service can have thousands of ports, and we assign an external IP for each LAN to access these services. This results in the required ipset size reaching "ports num" * "external IP num," which can exceed the default limit of 65536.When the ipset size reaches its limit, any new services added will not take effect.

By making these parameters configurable, users can adjust the size based on their specific needs, providing greater flexibility and adaptability. Additionally, when modifying `maxElem`, users should also consider adjusting `hashSize` to achieve optimal performance, allowing them to balance memory usage according to their current scenario.

This feature will enhance kube-proxy's scalability and performance in diverse environments.

There are also many users like us who need to expose a large number of ports：
https://github.com/kubernetes/kubernetes/issues/23864
https://github.com/kubernetes/kubernetes/issues/23864#issuecomment-373937794

**Comment by weak-fox** [comment]
@kubernetes/sig-network-feature-requests

**Comment by aroradaman** [comment]
The request seems reasonable
/triage accepted

**Comment by uablrek** [comment]
> 65536.When the ipset size reaches its limit, any new services added will not take effect.

What does "new services" mean here? Does overstepping the limit for one service prevent new ones?

I tried to recreate the problem with a *very big* service:
```yaml
apiVersion: v1
kind: Service
metadata:
  name: big-svc
spec:
  selector:
    app: tserver
  ports:
  - port: 5000
    name: p5000
  - port: 5001
    name: p5001
# ... 10000 ports goes here
  externalIPs:
  - 10.0.0.0
  - 10.0.0.1
# ... a number of ip's goes here
```
If I add 6 externalIPs it works (sort of), but if I make that 7, this shows up in the kube-proxy logs:

```
 > ipSetEntry="10.0.0.6,tcp:6656" ipSet="KUBE-EXTERNAL-IP"
E1228 09:05:11.483159     427 ipset.go:180] "Failed to add ip set entry to ip set" err=<
        error adding entry 10.0.0.0,tcp:10966, error: exit status 1 (ipset v6.38: Hash is full, cannot add more elements
```

**Comment by uablrek** [comment]
Is the "big-svc" above resembling what you use? (not considering UDP, which is irrelevant for the problem)

If so, you basically *must* switch to proxy-mode=nftables! proxy-mode=ipvs is putting an awful stress on the cluster, because it scales badly (not ipvs itself, but how kube-proxy uses it).

I tried the "big-svc" above with proxy-mode=nftables and with 7 externalIPs (10000x7=70000), and got no errors in the kube-proxy logs. Besides, the background cpu-load went down to ~0, compared to *a lot more* with ipvs. Just the time between create of the service until "ipvsadm" shows something sane is like several minutes, and the periodic sync with ipvs draws *a lot* of cpu.

With nftables it's ~10sec:
```
I1228 09:27:23.251618     439 proxier.go:1187] "SyncProxyRules complete" ipFamily="IPv4" elapsed="10.635019774s"
```
And the periodic sync just kicks in if needed (so ~0 cpu).

**Comment by weak-fox** [comment]
> > 65536.When the ipset size reaches its limit, any new services added will not take effect.
> 
> What does "new services" mean here? Does overstepping the limit for one service prevent new ones?
> 
> I tried to recreate the problem with a _very big_ service:
> 
> ```yaml
> apiVersion: v1
> kind: Service
> metadata:
>   name: big-svc
> spec:
>   selector:
>     app: tserver
>   ports:
>   - port: 5000
>     name: p5000
>   - port: 5001
>     name: p5001
> # ... 10000 ports goes here
>   externalIPs:
>   - 10.0.0.0
>   - 10.0.0.1
> # ... a number of ip's goes here
> ```
> 
> If I add 6 externalIPs it works (sort of), but if I make that 7, this shows up in the kube-proxy logs:
> 
> ```
>  > ipSetEntry="10.0.0.6,tcp:6656" ipSet="KUBE-EXTERNAL-IP"
> E1228 09:05:11.483159     427 ipset.go:180] "Failed to add ip set entry to ip set" err=<
>         error adding entry 10.0.0.0,tcp:10966, error: exit status 1 (ipset v6.38: Hash is full, cannot add more elements
> ```

I believe your attempt aligns with my scenario.When the rules are full, I create a new service. However, accessing it via the service name and port does not work.

**Comment by weak-fox** [comment]
> Is the "big-svc" above resembling what you use? (not considering UDP, which is irrelevant for the problem)
> 
> If so, you basically _must_ switch to proxy-mode=nftables! proxy-mode=ipvs is putting an awful stress on the cluster, because it scales badly (not ipvs itself, but how kube-proxy uses it).
> 
> I tried the "big-svc" above with proxy-mode=nftables and with 7 externalIPs (10000x7=70000), and got no errors in the kube-proxy logs. Besides, the background cpu-load went down to ~0, compared to _a lot more_ with ipvs. Just the time between create of the service until "ipvsadm" shows something sane is like several minutes, and the periodic sync with ipvs draws _a lot_ of cpu.
> 
> With nftables it's ~10sec:
> 
> ```
> I1228 09:27:23.251618     439 proxier.go:1187] "SyncProxyRules complete" ipFamily="IPv4" elapsed="10.635019774s"
> ```
> 
> And the periodic sync just kicks in if needed (so ~0 cpu).


Thanks for the test, it looks like switching proxy-mode to nftables in the latest version solves this problem for me.

I am currently using k8s 1.24.6. I checked that the relevant code does not contain nftables. I originally wanted to merge the code into the master and transplant it to a lower version, because I think that whether it is the latest version or a lower version, if we use ipvs will encounter this problem.

However, based on the response from this comment https://github.com/kubernetes/kubernetes/pull/129390#issuecomment-2563970958, I see that adding new features to this mode might encounter some difficulties. I feel we need to weigh the options: should we only provide it in the old version? not provide it at all? or consider other approaches?

Forgive me as I am new here. From my perspective, I believe this value should not be hardcoded and should be made available to users.

**Comment by uablrek** [comment]
> I believe this value should not be hardcoded and should be made available to users.

I agree, but proxy-mode=ipvs is a dead end. I tested proxy-mode=iptables, and it was much better than I feared.
```
I1228 12:18:54.254200     432 proxier.go:1547] "Reloading service iptables data" ipFamily="IPv4" numServices=10001 numEndpoints=20001 numFilterChains=5 numFilterRules=2 numNATChains=40004 numNATRules=170006
I1228 12:18:55.125366     432 proxier.go:822] "SyncProxyRules complete" ipFamily="IPv4" elapsed="1.005044042s"
```
The wopping 170006 NATRules loads *very* fast. I haven't tested traffic performance though, but unlike ipvs, iptables relies on conntrack only once the connection is setup. So you may try proxy-mode=iptables.

But don't try to list the iptables rules :smile:

**Comment by uablrek** [comment]
Disclaimer: proxy-mode=iptables has also been updated since v1.24 (which is old), so the performance may not be as good as in the latest release (I actually tested on "master").

Perhaps it's time to ask @danwinship for a comment. He wrote proxy-mode=nftables, and has optimized proxy-mode=iptables several times.

**Comment by weak-fox** [comment]
> Disclaimer: proxy-mode=iptables has also been updated since v1.24 (which is old), so the performance may not be as good as in the latest release (I actually tested on "master").

I will also test the rule matching efficiency of iptables and ipvs created by kube-proxy when there are a large number of rules.

> Perhaps it's time to ask @danwinship for a comment. He wrote proxy-mode=nftables, and has optimized proxy-mode=iptables several times.

Looking forward to Dan’s reply and  suggestions😁

**Comment by uablrek** [comment]
@weak-fox Since you can program, and it looks like you have some control of your clusters, please consider to write a custom proxy. I suggested that in:

https://github.com/kubernetes/enhancements/pull/2611#issuecomment-1059718812

Your use-case seem to be: load-balance external traffic via `externalIPs` _only_ to an all-ip service, which makes it simple (yes, actually simple). You would best use the `sigs.k8s.io/knftables` lib (same as kube-proxy), since `kpng` is archived.

Of course, Gateway-API is also an option. Perhaps a better one. But in-tree K8s support for all-ip services is extremely unlikely.

**Comment by weak-fox** [comment]
> Your use-case seem to be: load-balance external traffic via `externalIPs` _only_ to an all-ip service, which makes it simple (yes, actually simple). You would best use the `sigs.k8s.io/knftables` lib (same as kube-proxy), since `kpng` is archived.
> 
> Of course, Gateway-API is also an option. Perhaps a better one. But in-tree K8s support for all-ip services is extremely unlikely.

Yes, this may solve my problem. I will look into it.

**Comment by weak-fox** [comment]
>  I feel we need to weigh the options: should we only provide it in the old version? not provide it at all? or consider other approaches?
> 
> Forgive me as I am new here. From my perspective, I believe this value should not be hardcoded and should be made available to users.

Based on our discussion, I understand that the current Kubernetes features can meet my needs to some extent. To ensure the stability of the Kubernetes version, let's hold off on providing the ability to customize these two parameters for now.

But I can't shake the feeling that having these values hardcoded is a bit weird😅

If these challenges can be addressed, I believe adding these parameters has no obvious drawbacks and offers greater flexibility. I am also willing to try and resolve these issues❤️

**Comment by weak-fox** [comment]
> The wopping 170006 NATRules loads _very_ fast. I haven't tested traffic performance though, but unlike ipvs, iptables relies on conntrack only once the connection is setup. So you may try proxy-mode=iptables.

@uablrek 
In k8s 1.24.6, when I try to switch to iptables mode and created 20 IPs (each with 10000 ports) , the `iptables-restore` command failed and caused extremely high CPU utilization. The CPU usage reached 99.38% across our 72-core cluster (consisting of 3 master nodes and 6 worker nodes). 

```shell
E1230 06:22:11.668855       1 proxier.go:1481] "Failed to execute iptables-restore" err=<
        exit status 1: iptables-restore: line 220139 failed
 >
I1230 06:22:11.668871       1 proxier.go:858] "Sync failed" retryingTime="30s"
```

Then I tried to create 7 IPs (each IP has 10,000 ports), and I noticed that when I run iptables-save, there are approximately 250,000 lines.

**Comment by uablrek** [comment]
:cry: Guess https://github.com/kubernetes/kubernetes/issues/129389#issuecomment-2564315937 was motivated.

If I had this problem, I would go for a custom proxy with _one_ use-case: yours. It becomes sooo much simpler if you can ditch all "generic" stuff. Further, I think your use-case, all-ip _only_ for external traffic, is the most common. So, if you do it you may get a community around it. I'll be happy to contribute, but I am basically retired now, so I will not "drive" a project, and I code just for fun :smile:

**Comment by aojea** [comment]
> the iptables-restore command failed and caused extremely high CPU utilization.

the large scale optimization for kube-proxy was implemented in 1.28 https://github.com/kubernetes/kubernetes/pull/117787

It is important to mention that it also depends a lot on the userspace libraries, check what iptables userspace versions you are using ... and the kernel backends, if is the legacy or nft https://developers.redhat.com/blog/2020/04/27/optimizing-iptables-nft-large-ruleset-performance-in-user-space#baseline_results

> I am currently using k8s 1.24.6. I checked that the relevant code does not contain nftables. I originally wanted to merge the code into the master and transplant it to a lower version, because I think that whether it is the latest version or a lower version, if we use ipvs will encounter this problem.

features are not backported, so you can only use the feature since the version it was added, there is a high bar for cherrypicks to avoid having regressions on stable branches https://github.com/kubernetes/community/blob/master/contributors/devel/sig-release/cherry-picks.md#what-kind-of-prs-are-good-for-cherry-picks

**Comment by danwinship** [comment]
So, _if_ we were going to do something here, it seems to me that the right fix would be for the IPVS backend to notice that it's going to overflow the sets, and create new ones in that case.

But that would be a big change, and as others have noted, the ipvs code is not entirely maintained (in that nobody cares when its e2e tests fail), and is semi-deprecated in favor of nftables. And the reporter only wants the feature in an old release, _and_ it's a weird use case that is probably better served by Gateway API.

So maybe WONTFIX?

**Comment by danwinship** [comment]
> But in-tree K8s support for all-ip services is extremely unlikely.

I feel like the big problem with all-IP services is that it was never clear that it was actually a single feature, as opposed to several very similar sounding features with different requirements. (IIRC, the KEP proposed that it would be for ClusterIP services only, but many of the people "+1"-ing the cared _only_ about the LoadBalancer/ExternalIP case.)

**Comment by uablrek** [comment]
Intra-cluster all-IP applications can use headless services. Actually, from querying SIP developers, that's the preferred way (well at least it was 2-3 years back). I don't know about RTP, but _if_ an app needs to stream RTP intra-cluster, perhaps headless services can be used.

**Comment by weak-fox** [comment]
> It is important to mention that it also depends a lot on the userspace libraries, check what iptables userspace versions you are using

It's `iptables v1.8.7 (nf_tables)`. 

> features are not backported, so you can only use the feature since the version it was added, there is a high bar for cherrypicks to avoid having regressions on stable branches 

I will try to upgrade the kube-proxy version and use nftables to see if there are any issues. Do you have any suggestions or recommendations regarding version upgrade considerations or specific checks when upgrading only kube-proxy?

**Comment by weak-fox** [comment]
> but many of the people "+1"-ing the cared _only_ about the LoadBalancer/ExternalIP case

Yes, my example can also be abstracted as accessing all ports of services from outside the cluster, for instance: through ExternalIP.

**Comment by weak-fox** [comment]
> I don't know about RTP, but _if_ an app needs to stream RTP intra-cluster, perhaps headless services can be used.

Headless services can solve internal access issues within a cluster. However, currently the devices are outside the cluster and use standard SIP/RTP protocols for video transmission.

It appears there are only two solutions available:
1. Use custom proxy
2. Upgrade to use a newer version of kube-proxy and use nftables

**Comment by danwinship** [comment]
Yes. There is no fully-baked solution for doing SIP/RTP within Kubernetes. I think lots of people would be happy if there was one, but nobody has ever stepped up to try to figure out that solution. There was a long discussion about supporting port ranges, where it's clear that most of the people involved wanted to do SIP/RTP, but as I was saying about, there were contradictory requests (probably in part because some of the people in the discussion had some use case other than SIP/RTP in mind).

Basically, people come up with hacky ways of doing SIP/RTP in Kubernetes using the existing features, and then they try to propose small features that would make their existing hacks a little bit less hacky. But we don't want that. We want a _good_ solution for SIP/RTP in Kubernetes. Someone needs to start from the top level, and explain, here are the detailed user stories of how people want SIP/RTP in Kubernetes to work, here is the architecture for how this would be deployed, and here are the parts of that that there aren't currently good ways to implement. Then we can figure out what the right _non_-hacky features are that we need.

OTOH, it might be the case that the right answer here (especially for the case of SIP/RTP with cluster-external clients) is a SIP/RTP-specific Gateway implementation, in the same way that https://github.com/kubernetes-sigs/gateway-api-inference-extension is an LLM-specific Gateway implementation. That wouldn't need a KEP, it would just need someone to write it. (Though it could still potentially benefit from a KEP or WG or something to make sure it wasn't limited to too specific a use case.)

**Comment by weak-fox** [comment]
Thank you all for your responses. I'm not particularly knowledgeable about SIP/RTP myself; my involvement comes from helping clients address these issues. These requirements indeed don't form a complete picture and may not always be reasonable (for example, many devices were designed with only local networks in mind, not the cloud). 

Through the discussions and attempts during this period, I feel that we need more knowledgeable individuals to help establish standards through ongoing discussions. It's not just about solving a specific requirement unless that requirement can be abstracted into a capability that Kubernetes truly needs. 

Regarding the external access to multiple ports, I don't have many good ideas at the moment. I'll wait for about 7 days to see if there are any more responses, and if not, I'll close this issue.

**Comment by weak-fox** [comment]
At present, it appears there isn't a perfect solution available. Kubernetes has also encountered some issues with IPVS testing. In my scenario, I tried the modifications from this PR https://github.com/kubernetes/kubernetes/pull/129390, using maxElements set to 160,000, and haven't experienced any issues so far. I'm closing this issue, but if others encounter similar problems without finding a good solution, they might want to try the modifications from this PR as well.
